tests/host_tools/fcmetrics.py

```diff
@@ -363,3 +367,126 @@ def validate_per_device_metrics(self, fc_metrics):
+        "latencies_us.full_create_snapshot",
+        "latencies_us.diff_create_snapshot",
+        "latencies_us.load_snapshot",
+        "latencies_us.pause_vm",
+        "latencies_us.resume_vm",
+        "latencies_us.vmm_full_create_snapshot",
+        "latencies_us.vmm_diff_create_snapshot",
+        "latencies_us.vmm_load_snapshot",
+        "latencies_us.vmm_pause_vm",
+        "latencies_us.vmm_resume_vm",
+    }
+    skip = set()
+    for group, keys in fc_metrics.items():
+        if group == "utc_timestamp_ms":
+            continue
+        for key, value in keys.items():
+            full_key = f"{group}.{key}"
+            # values are 0 when:
+            # - there is no update
+            # - device is not used
+            # - SharedIncMetric reset to 0 on flush so if
+            #   there is no change metric the values remain 0.
+            # We can save the amount of bytes we export to
+            # CloudWatch in these cases.
+            # The other cases of 0 are failures in which case
+            # we should assert instead of sending to CW.
+            if (
+                "fail" in full_key
+                or "err" in full_key
+                or "num_faults" in full_key
+                or "panic" in full_key
+            ) and assert_on_err is True:
+                assert 0 == value, f"Error reported in metrics: {full_key}={value}"
+
+            if value != 0:
+                if full_key in skip:
+                    continue
+                unit = get_emf_unit_for_fc_metrics(key)
+                metrics.put_metric(f"fc_metrics.{full_key}", value, unit=unit)
+                # print(f"fc_metrics.{full_key} {value}, unit={unit}")
+                if full_key in metrics_to_export_once:
+                    skip.add(full_key)
+
+
+class FCMetricsMonitor(Thread):
+    """
+    read Firecracker metrics from the microvm every 60secs and
+    uploads the metrics to CW.
+    We do this as a daemon thread every 60sec, instead of collecting
+    all metrics together in the end, to retain timestamp of the metrics.
+    """
+
+    def __init__(self, vm, dimensions, assert_on_err=False):
+        Thread.__init__(self, daemon=True)
+        self.vm = vm
+        self.dimensions = dimensions
+        self.assert_on_err = assert_on_err
+
+        self.metrics_index = 0
+        self.running = True
+
+        self.metrics_logger = get_metrics_logger()
+        self.metrics_logger.set_dimensions(dimensions)
+
+    def _flush_metrics(self):
+        """
+        Since vm.flush_metrics provides only the latest metrics,
+        we call vm.get_all_metrics() instead to be able to collect
+        and upload all metrics data emitted by the microvm.
+        This utility function is created to keep common code in one
+        place and is called every 60sec once the daemon starts and then
+        once when the daemon stops.
+        """
+        metrics = self.vm.get_all_metrics()
```

Comment:
Since we poll every 60s, then flush_metrics should be enough? If we want to have a separate, one-shot method that's ok (though we don't have a use for it in the PR), but I think having both here makes this more complex.

```
def _flush_metrics(self):
    flush_fc_metrics_to_cw(self.vm.flush_metrics())

def flush_metrics_all(self):
    for metrics in self.vm.get_all_metrics():
        flush_fc_metrics_to_cw(metrics)
```

Reply:
I find get_all_metrics() more safe because flush_metrics() there is a chance to miss some metrics entries.
e.g. suppose there are some incmetrics like block read_bytes just flushed by FC and they are the last entry in fcmetrics ndjson file, now if we call vm.flush_metrics() it will send a new api request to flush FC metrics and latest entry of fcmetrics ndjson file would now point to some different read_bytes data. We lose the previously filled read_bytes metrics in these cases so calling get_all_metrics() and iterating over all it seemed safe to capture all metrics.

Reply:
Thanks for the explanation. I guess we can always optimize it later if need be.